runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -10412,227 +10412,767 @@ static TR::Register *inlineIsAssignableFrom(TR::Node *node, TR::CodeGenerator *c
+         third4Reg = cg->allocateRegister(TR_VRF);
+         fourth4Reg = cg->allocateRegister(TR_VRF);
+         // fall through
+      case TR::Int16:
+         high4Reg = cg->allocateRegister(TR_VRF);
+         if (!isSigned)
+            vunpackMaskReg = cg->allocateRegister(TR_VRF); // for masking unsigned types
+         // fall through
+      case TR::Int32:
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+
+   // the hashReg is zero initially, unless specified otherwise
+   TR::Register *hashReg = NULL;
+   if (nonZeroInitial)
+      {
+      hashReg = cg->gprClobberEvaluate(hashNode);
+      }
+   else
+      {
+      hashReg = cg->allocateRegister();
+      generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, hashReg, 0x0);
+      }
+
+   TR::LabelSymbol *serialLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *serialLoopLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *serialUnrollLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *special2Label = generateLabelSymbol(cg); // for the length-2 special case
+   TR::LabelSymbol *VSXLoopPrep = generateLabelSymbol(cg);
+   TR::LabelSymbol *VSXLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *POSTVSXLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *endLabel = generateLabelSymbol(cg);
+
+   // Skip header of the array
+   intptr_t hdrSize = TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, valueReg, valueReg, hdrSize);
+
+   // v = v + offset<<lg(elementSize)
+   // end = v + count<<lg(elementSize)
+   switch (elementType)
+      {
+      case TR::Int8:
+         break;
+      case TR::Int16:
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, endReg, endReg, endReg);
+         generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, vendReg, vendReg, vendReg);
+         break;
+      case TR::Int32:
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldicr, node, endReg, endReg, 2,
+                                         CONSTANT64(0xFFFFFFFFFFFFFFFC));
+         generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldicr, node, vendReg, vendReg, 2,
+                                         CONSTANT64(0xFFFFFFFFFFFFFFFC));
+         break;
+      default:
+         TR_ASSERT_FATAL(false, "Unsupported hashCodeHelper elementType");
+      }
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, valueReg, valueReg, endReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, endReg, valueReg, vendReg);
+
+   // initialisations
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, tempReg, 0x0);
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, constant0Reg, 0x0);
+
+   // using the serial loop is faster if there are less than 16 items
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpi4, node, condReg, vendReg, 16*elementSize);
```

Comment:
i naturally have questions re this condition if it is optimal (verified) for those different scenarios (byte/char/short/int).